In [ ]:
%run 'Surprise Housing - Part 1.ipynb'

### **Question 1**

- What is the optimal value of alpha for ridge_doubled_doubled_doubled_doubled and lasso regression?

In [6]:
# Optimal value of alpha for ridge
print('In Ridge Regression, the optimal value of alpha is', model_cv.best_params_['alpha'])

# Optimal value of alphe for lasso
print('In Lasso Regression, the optimal value of alpha is', model_cv_lasso.best_params_['alpha'])

In Ridge Regression, the optimal value of alpha is 200
In Lasso Regression, the optimal value of alpha is 500


- What will be the changes in the model if you choose double the value of alpha for both ridge and lasso?

In [7]:
ralpha = 400
ridge_doubled = Ridge(alpha=ralpha)
ridge_doubled.fit(X_train, y_train)

y_pred_ridge_doubled_train = ridge_doubled.predict(X_train)
y_pred_ridge_doubled_test = ridge_doubled.predict(X_test)

metric2_doubled = []
r2_train_lr = r2_score(y_train, y_pred_ridge_doubled_train)
metric2_doubled.append(r2_train_lr)

r2_test_lr = r2_score(y_test, y_pred_ridge_doubled_test)
metric2_doubled.append(r2_test_lr)

rss1_lr = np.sum(np.square(y_train - y_pred_ridge_doubled_train))
metric2_doubled.append(rss1_lr)

rss2_lr = np.sum(np.square(y_test - y_pred_ridge_doubled_test))
metric2_doubled.append(rss2_lr)

mse_train_lr = mean_squared_error(y_train, y_pred_ridge_doubled_train)
metric2_doubled.append(mse_train_lr**0.5)

mse_test_lr = mean_squared_error(y_test, y_pred_ridge_doubled_test)
metric2_doubled.append(mse_test_lr**0.5)

In [34]:
lalpha = 1000
lasso_doubled = Lasso(alpha=lalpha)
lasso_doubled.fit(X_train, y_train)
y_pred_lasso_doubled_train = lasso_doubled.predict(X_train)
y_pred_lasso_doubled_test = lasso_doubled.predict(X_test)

metric3_doubled = []
r2_train_lr = r2_score(y_train, y_pred_lasso_doubled_train)
metric3_doubled.append(r2_train_lr)

r2_test_lr = r2_score(y_test, y_pred_lasso_doubled_test)
metric3_doubled.append(r2_test_lr)

rss1_lr = np.sum(np.square(y_train - y_pred_lasso_doubled_train))
metric3_doubled.append(rss1_lr)

rss2_lr = np.sum(np.square(y_test - y_pred_lasso_doubled_test))
metric3_doubled.append(rss2_lr)

mse_train_lr = mean_squared_error(y_train, y_pred_lasso_doubled_train)
metric3_doubled.append(mse_train_lr**0.5)

mse_test_lr = mean_squared_error(y_test, y_pred_lasso_doubled_test)
metric3_doubled.append(mse_test_lr**0.5)

- What will be the most important predictor variables after the change is implemented?

In [35]:
# [Ridge] Original list of most important predictors (before doubling alpha)
betas['Ridge'].sort_values(ascending=False).head(5)

GrLivArea               10534.730475
OverallQual             10450.014545
2ndFlrSF                 7347.009310
RoofMatl_WdShngl         7231.974776
Neighborhood_NoRidge     6892.484145
Name: Ridge, dtype: float64

In [36]:
# Creating a table with doubled values
betas_doubled = pd.DataFrame(index=X.columns)
betas_doubled.rows = X.columns
betas_doubled['Linear'] = lm.coef_
betas_doubled['Ridge_Doubled'] = ridge_doubled.coef_
betas_doubled['Lasso_Doubled'] = lasso_doubled.coef_

#pd.set_option('display.max_rows', None)
#betas_doubled.head(80)

In [37]:
# [Ridge Doubled] List of most important predictors after doubling alpha
betas_doubled['Ridge_Doubled'].sort_values(ascending=False).head(5)

OverallQual             9146.386678
GrLivArea               8691.378921
Neighborhood_NoRidge    6361.309083
Neighborhood_NridgHt    5674.340025
1stFlrSF                5618.975207
Name: Ridge_Doubled, dtype: float64

##### **Answer - Part 1:**
- In Ridge, `OverallQual` and `GrLivArea` switched top two places when alpha was doubled.
- `Neighborhood_NoRidge` came 3rd after being 5th before doubling alpha, followed by `Neighborhood_NridgHt`, which indicates that the location contributes a lot in the price.
- `1stFlrSF` has taken over `2ndFlrSF`, indicating that price is mostly decided by the main floor area.

In [38]:
# [Lasso] Original list of most important predictors (before doubling alpha)
betas['Lasso'].sort_values(ascending=False).head(5)

RoofMatl_CompShg    32722.709855
GrLivArea           29737.474373
RoofMatl_WdShngl    20224.204921
RoofMatl_Tar&Grv    20030.873653
RoofMatl_WdShake    13766.132756
Name: Lasso, dtype: float64

In [39]:
# [Lasso Doubled] Original list of most important predictors (before doubling alpha)
betas_doubled['Lasso_Doubled'].sort_values(ascending=False).head(5)

GrLivArea               27139.877611
OverallQual             16569.350649
Neighborhood_NridgHt     7429.544401
Neighborhood_NoRidge     7142.238110
GarageCars               6811.531363
Name: Lasso_Doubled, dtype: float64

##### **Answer - Part 2:**
- In Lasso, all roof materials types were dropped from top features.
- `GrLivArea` came 1st after being 2nd before doubling alpha, followed by `OverallQual`.
- `lasso` and `ridge` has predicted the same top 4 variables, meaning that the new (doubled) alpha values made both models the same.
- `GarageCars` came in the fifth place. It was one of the highest correlated values to `SalePrice` before running regression.

### **Question 2**

You have determined the optimal value of lambda for ridge and lasso regression during the assignment.
- Now, which one will you choose to apply and why?


In [40]:
# Features selected by Lasso
features = betas[betas['Lasso'] != 0]
features_in_double = betas_doubled[betas_doubled['Lasso_Doubled'] != 0]
print('Total features are', betas.shape[0])
print('Lasso has selected', betas.shape[0] - features.shape[0], 'and dropped',
       features.shape[0], 'features before doubling alpha')
print('Lasso has selected', betas_doubled.shape[0] - features_in_double.shape[0], 'and dropped', 
      features_in_double.shape[0], 'features after doubling alpha')

Total features are 246
Lasso has selected 118 and dropped 128 features before doubling alpha
Lasso has selected 154 and dropped 92 features after doubling alpha


In [41]:
# Creating a table which contain all the metrics

lr_table_doubled = {'Metric': ['R2 Score (Train)','R2 Score (Test)','RSS (Train)','RSS (Test)',
                       'MSE (Train)','MSE (Test)'], 
        'Linear Regression': metric
        }

lr_metric_doubled = pd.DataFrame(lr_table_doubled ,columns = ['Metric', 'Linear Regression'] )

rg_metric_doubled = pd.Series(metric2_doubled, name = 'Ridge Regression')
ls_metric_doubled = pd.Series(metric3_doubled, name = 'Lasso Regression')

final_metric_doubled = pd.concat([lr_metric_doubled, rg_metric_doubled, ls_metric_doubled], axis = 1)

In [46]:
# Original metrics against alpha_ridge = 200 and alpha_lasso = 500
final_metric.head(6)

Metric  Linear Regression  Ridge Regression  Lasso Regression
0  R2 Score (Train)       9.492998e-01      9.042479e-01      9.237615e-01
1   R2 Score (Test)      -1.608086e+21      8.591934e-01      8.444495e-01
2       RSS (Train)       3.235028e+11      6.109654e+11      4.864548e+11
3        RSS (Test)       4.532733e+33      3.968933e+11      4.384521e+11
4       MSE (Train)       1.780025e+04      2.446220e+04      2.182772e+04
5        MSE (Test)       3.216940e+15      3.010231e+04      3.163909e+04

In [47]:
# Metrics after doubling alpha value for both ridge and alpha
final_metric_doubled.head(6)

Metric  Linear Regression  Ridge Regression  Lasso Regression
0  R2 Score (Train)       9.492998e-01      8.898002e-01      8.923762e-01
1   R2 Score (Test)      -1.608086e+21      8.579731e-01      8.411681e-01
2       RSS (Train)       3.235028e+11      7.031520e+11      6.867151e+11
3        RSS (Test)       4.532733e+33      4.003329e+11      4.477014e+11
4       MSE (Train)       1.780025e+04      2.624289e+04      2.593435e+04
5        MSE (Test)       3.216940e+15      3.023247e+04      3.197107e+04

##### **Answer:** 
- Ridge regression has higher R2 score, and lower gap between train and test sets.
- Doubling alpha lowered R2 score by average of 2-3% and increased MSE in both models, which is not helpful.
- Lasso's features selection is useful in case of large amount of variables. Doubling alpha also showed how Lasso has performed by dropping less features.
- On the other hand, with lower MSE on Test set in Ridge `3.01` vs `3.16` in Lasso, this implies that Ridge performed better on unseen data.

Lasso is preferred in this case study due to its ability to drop features that are less significant to `SalePrice`, and this is effective since we have a lot of variables (246) in this case.

### **Question 3**

After building the model, you realised that the five most important predictor variables in the lasso model are not available in the incoming data. You will now have to create another model excluding the five most important predictor variables. Which are the five most important predictor variables now?


In [48]:
lasso_top5 = betas['Lasso'].sort_values(ascending=False).head(5)
lasso_top5

RoofMatl_CompShg    32722.709855
GrLivArea           29737.474373
RoofMatl_WdShngl    20224.204921
RoofMatl_Tar&Grv    20030.873653
RoofMatl_WdShake    13766.132756
Name: Lasso, dtype: float64

In [49]:
X_new = X.drop(columns=['GrLivArea', 'RoofMatl_WdShngl', 'RoofMatl_Tar&Grv', 'RoofMatl_WdShake', 'RoofMatl_CompShg'], axis=1)

# split into train and test
from sklearn.model_selection import train_test_split
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y, 
                                                    train_size=0.7,
                                                    test_size = 0.3, random_state=100)

In [50]:
lasso = Lasso()
params = {'alpha': [0.0001, 0.001, 0.01, 0.1, 0.2, 0.5,
 1.0, 2.0, 5.0, 10.0, 20, 50, 100, 200, 500, 1000]}

# cross validation
model_cv_lasso_new = GridSearchCV(estimator = lasso, 
                        param_grid = params, 
                        scoring= 'neg_mean_absolute_error', 
                        cv = folds, 
                        return_train_score=True,
                        verbose = 1)            

model_cv_lasso_new.fit(X_train_new, y_train_new) 

Fitting 5 folds for each of 16 candidates, totalling 80 fits


GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': [0.0001, 0.001, 0.01, 0.1, 0.2, 0.5, 1.0, 2.0,
                                   5.0, 10.0, 20, 50, 100, 200, 500, 1000]},
             return_train_score=True, scoring='neg_mean_absolute_error',
             verbose=1)

In [51]:
print(model_cv_lasso_new.best_params_)

{'alpha': 500}


In [52]:
alpha = model_cv_lasso_new.best_params_['alpha']
lasso_new = Lasso(alpha=alpha)

lasso_new.fit(X_train_new, y_train_new)

y_pred_lasso_train_new = lasso_new.predict(X_train_new)
y_pred_lasso_test_new = lasso_new.predict(X_test_new)

In [53]:
# Creating a table with new values
betas_new = pd.DataFrame(index=X_new.columns)
betas_new.rows = X_new.columns
betas_new['Lasso_New'] = lasso_new.coef_

##### **Answer:** New top 5 features now are:

In [54]:
betas_new['Lasso_New'].sort_values(ascending=False).head(5)

2ndFlrSF                24211.243146
1stFlrSF                18551.565585
OverallQual             15063.322878
GarageCars               7856.510694
Neighborhood_NoRidge     7737.314025
Name: Lasso_New, dtype: float64

### **Question 4**

- How can you make sure that a model is robust and generalisable?

    **Answer:** 
    - When the model performs well on test data compared to an equivalent training data, it means that it's generalised well. This means also the model did not over or under fit the data.
    - On the other hand, for a model to be robust, it should perform well on test (unseen) data when training data is noisy (e.g.: missing/null values), or has outliers or variations.

##
- What are the implications of the same for the accuracy of the model? and why?

    **Answer:**
    - Robustness (as defined above) goes opposite to accuracy, as for the model to be robust, it tends to be more complex, leading to lower bias, higher variance, and hence too much accuracy.
    - There's always a tradeoff between bias and variance. To reach an optimal model, we'll sacrifice some robustness towards optimally predicting values that were not used to train the model, which is the measurement of the accuracy.
##

#### Note:

- In this case study, we cleaned, prepared and scaled the data before splitting into training and test sets, so that both are equivalent. So our model can easily be considered to be well generalised. Below values of accuracy proves this point.
- If we wanted to make the model more robust, we could've split the data before cleaning, then train the model. In this case, the test set would include lots of noise (missing values). Then apply regularisation and testing different alpha values for both ridge and lasso.   

In [55]:
def MAPE(Y_actual,Y_Predicted):
    mape = np.mean(np.abs((Y_actual - Y_Predicted)/Y_actual))*100
    return mape

Ridge_MAPE = MAPE(y_test, y_pred_ridge_test)
Ridge_Accuracy = 100 - Ridge_MAPE
print('Accuracy of Ridge Regression: {:0.2f}%.'.format(Ridge_Accuracy))
print('')
Lasso_MAPE = MAPE(y_test,y_pred_lasso_test)
Lasso_Accuracy = 100 - Lasso_MAPE
print('Accuracy of Lasso Regression: {:0.2f}%.'.format(Lasso_Accuracy))


Accuracy of Ridge Regression: 89.40%.

Accuracy of Lasso Regression: 89.70%.
